unblob/handlers/filesystem/netgearchk.py

```diff
@@ -0,0 +1,76 @@
+import io
+from pathlib import Path
+from typing import Optional
+
+from structlog import get_logger
+
+from unblob.extractor import carve_chunk_to_file
+from unblob.file_utils import Endian, File, StructParser
+from unblob.models import Chunk, Extractor, HexString, StructHandler, ValidChunk
+
+logger = get_logger()
+
+CHK_HEADER = r"""
+        typedef struct chk_header {
+            uint32 magic;           /* "2A 23 24 5E" */
+            uint32 header_len;
+            uint8  reserved[8];
+            uint32 kernel_chksum;
+            uint32 rootfs_chksum;
+            uint32 kernel_len;
+            uint32 rootfs_len;
+            uint32 image_chksum;
+            uint32 header_chksum;
+            /* char board_id[total_header_len - header_len] upto MAX_BOARD_ID_LEN */
+            /* uint32 total_header_len; */
+        } chk_header_t;
+    """
+
+
+class CHKExtractor(Extractor):
+    def __init__(self):
+        self._struct_parser = StructParser(CHK_HEADER)
+
+    def extract(self, inpath: Path, outdir: Path):
+        with File.from_path(inpath) as file:
+            header = self._struct_parser.parse("chk_header_t", file, Endian.BIG)
+
+            file.seek(header.header_len, io.SEEK_SET)
+
+            self._dump_file(file, outdir, Path("kernel"), header.kernel_len)
+            self._dump_file(file, outdir, Path("rootfs"), header.rootfs_len)
+
+    def _dump_file(self, file: File, outdir: Path, path: Path, length: int):
+        if not length:
+            return
+
+        start = file.tell()
+        chunk = Chunk(start_offset=start, end_offset=start + length)
+        carve_chunk_to_file(outdir.joinpath(path), file, chunk)
+
+
+class NetgearCHKHandler(StructHandler):
+    NAME = "chk"
+
+    PATTERNS = [HexString("2a 23 24 5e")]
+
+    C_DEFINITIONS = CHK_HEADER
+    HEADER_STRUCT = "chk_header_t"
+    EXTRACTOR = CHKExtractor()
+
+    def calculate_chunk(self, file: File, start_offset: int) -> Optional[ValidChunk]:
+        header = self.parse_header(file, endian=Endian.BIG)
```

Comment:
We usually recommend to add a header validation function to make sure we don't end up with false positives. A simple implementation woud look like this:

```python
from dissect.cstruct import Instance
# --snip---
def is_valid_header(header: Instance) -> bool:
    if header.header_len != len(header):
        return False
     return True
```

Let's look at it together tomorrow so we expand the validation.

Reply:
Implemented the `is_valid_header` check aswell as a try expect to see if the `board_id` contained any character that were not `utf-8`